feat(mobile): thumbhash on mobile

shenlong-tanwen opened the pull request:
Continues on #2913 

#### Changes made in the PR

- Caught up to main and updated asset grid view to use thumbhash on scroll as well and not just when rendering the image
- Wraps `ThumbnailImage` in asset grid view with `RepaintBoundary` to prevent repaint of other assets in the same row due to repaint of a different asset

> [!Note]
> I did not test it extensively and not sure about the performance implications of loading the assets even during scrolling. The flickering is not completely removed though, but it is vastly reduced in my initial level of testing

lukehmcc committed:
add thumbhash support to mobile app

lukehmcc committed:
Merge branch 'main' of https://github.com/immich-app/immich into thumbhash_mobile

lukehmcc committed:
update generated files

lukehmcc committed:
fix tests

lukehmcc committed:
update migration and remove isar files

lukehmcc committed:
add migration and update gitignore to ignore isar

lukehmcc committed:
update canUpdate to take into account thumbhash

lukehmcc committed:
change thumbhash to List<byte> from String

lukehmcc committed:
Merge branch 'main' of https://github.com/immich-app/immich into thumbhash_mobile

lukehmcc committed:
remove needless imports

alextran1502 committed:
Merge branch 'main' of github.com:immich-app/immich into pr/lukehmcc/2913

alextran1502 committed:
Merge branch 'main' of github.com:immich-app/immich into pr/lukehmcc/2913

lukehmcc committed:
change fade in to ImageFade library

lukehmcc committed:
remove unnecesary import

lukehmcc committed:
add back proper error handling

lukehmcc committed:
add missing trailing comma

Commit:
Merge branch 'main' of github.com:immich-app/immich into pr/lukehmcc/2913

alextran1502 committed:
Merge branch 'main' of github.com:immich-app/immich into pr/lukehmcc/2913

lukehmcc committed:
merged thumbhash_mobile with upstream

shenlong-tanwen committed:
chore: pull main

shenlong-tanwen committed:
mobile: show thumbhash on scroll

shenlong-tanwen committed:
mobile: wrap thumbnail image with renderboundary

fyfrey commented on mobile/lib/modules/home/ui/asset_grid/immich_asset_grid_view.dart:
```diff
@@ -254,9 +264,8 @@ class ImmichAssetGridViewState extends State<ImmichAssetGridView> {
-        final List<Asset> assetsToRender = scrolling
-            ? []
-            : widget.renderList.loadAssets(section.offset, section.count);
+        final List<Asset> assetsToRender =
```
This loads many assets quickly and often from DB while scrolling. We'd need to test the performance implications for large libraries (500k assets) on medium-powered devices.

shenlong-tanwen replied:
Yes. This does load the assets for each of the section during scrolling. I gave it a test only in the simulator using the demo instance which is not much. This change is to also show the thumb hash on scrolling with the scroll bar.

The main change that vastly reduced the flickering was using the `RepaintBoundary` around `ThumbnailImage`. I tested it after removing all the changes I made from `immich_asset_grid_view` and except wrapping the thumbnails with repaintboundary.

We can remove the thumbhash on scroll change if the performance is bad on low - medium end devices. I don't have that many assets to test 😅  Would be great if someone could test it

alextran1502 committed:
Merge branch 'main' of github.com:immich-app/immich into feat/mobile-thumbhash-scroll

alextran1502 commented:
Hello, I think this works a lot better now. One scenario I noticed is that if we keep holding the thumb scroll and not release it, it will keep the section rebuilt. I think we can optimize it by disable thumbhash loading unless the thumb scroll is released

shenlong-tanwen commented:
> Hello, I think this works a lot better now. One scenario I noticed is that if we keep holding the thumb scroll and not release it, it will keep the section rebuilt. I think we can optimize it by disabling thumbhash loading unless the thumb scroll is released.
> 
> I think it will also reduce the "jitter" when dragging the thumb scroll in a large album view

Yes, we can do that. Fynn also had concerns regarding the performance of loading the assets from db on scroll which is needed to display the thumbhash on thumb scroll. Should I remove the scrolling behavior with thumb scroll back to displaying gray placeholders?

shenlong-tanwen committed:
mobile: use grey placeholders on thumb scroll

alextran1502 commented:
> > Hello, I think this works a lot better now. One scenario I noticed is that if we keep holding the thumb scroll and not release it, it will keep the section rebuilt. I think we can optimize it by disabling thumbhash loading unless the thumb scroll is released.
> > I think it will also reduce the "jitter" when dragging the thumb scroll in a large album view
> 
> Yes, we can do that. Fynn also had concerns regarding the performance of loading the assets from db on scroll which is needed to display the thumbhash on thumb scroll. Should I remove the scrolling behavior with thumb scroll back to displaying gray placeholders?

Yes, I think this would be preferred. We don't want to sacrifice performance for the aesthetic look in this use case 😄 .

shenlong-tanwen commented:
> Yes, I think this would be preferred. We don't want to sacrifice performance for the aesthetic look in this use case 😄 .

Updated my change now to only add a `RepaintBoundary` around the `ThumbnailImage` and removed the thumbhash on scroll 😅

shenlong-tanwen committed:
chore: pull main

shenlong-tanwen committed:
remove syncDuration

shenlong-tanwen committed:
chore: pull main

shenlong-tanwen commented:
Will re-implement this on top of #7016
